### ISSUE TITLE
Feature request: Ability to remove in-memory sessions and already sent net requests to free up memory.

### ISSUE THREAD

**Issue by 1nfility** [electron/electron#25631]
### Problem Description
I'm using the net request module, and in-memory sessions for sending requests. Sending many requests, and using many in-memory sessions can easily take up many hundreds of megabyte of memory. You are not able to free this memory up unless you restart your Electron process.

### Proposed Solution
I'd like to propose a solution, where you would be able to delete/remove an in-memory session, as well as already sent net requests.

**Comment by nornagon** [comment]
Can you say a little more about your use case? Net requests should not take up any memory after they're released. Sessions do stay around forever, so if you're creating a new session for every request, that could certainly result in increased memory usage. For most things there are ways to reuse a session, though, so you should be able to work with a small number of sessions for most things.

**Comment by 1nfility** [comment]
Yes. I'm using electron version 10.1.2. This example is by sending 600 requests at the same time to test each proxy.
As you can see, the memory goes from 150MB to 335MB, and is never released. In order to get the memory back down, you must restart the electron app. This is only 1 ipc call to the main process to test the 600 proxies.

This is *only* 600 net requests and 600 sessions in total.

**Process name**&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;**Memory**&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;**Threads**
<img width="759" alt="Skærmbillede 2020-09-29 kl  11 34 41" src="https://user-images.githubusercontent.com/40140509/94541544-fe75d200-0247-11eb-929b-2a80bf2721a5.png">
<img width="748" alt="Skærmbillede 2020-09-29 kl  11 35 11" src="https://user-images.githubusercontent.com/40140509/94541550-003f9580-0248-11eb-8a65-d6a428b31ed3.png">

**Comment by nornagon** [comment]
Can you reuse sessions?

**Comment by 1nfility** [comment]
We use sessions as a cookie jar and to use proxies in the requests. We could possibly re-use sessions, but a lot of memory is used and never released when making requests.

I don't know how 600 sessions and 600 requests can take up 157MB. In this example each session doesn't even have any cookies.

**Comment by nornagon** [comment]
@1nfility as I mentioned earlier, net requests shouldn't be leaking memory—if you have a test case which shows leaking memory with many net requests, but a **fixed number of sessions**, that's definitely something we can look into and fix.

You can use e.g. [`ses.clearStorageData()`](https://www.electronjs.org/docs/api/session#sesclearstoragedataoptions) to clear a session of its cookies and reuse it. Proxy settings are updatable between requests, too.

I agree that it would be convenient to be able to mint a fresh session for each request, but due to Chromium implementation details that tie a BrowserContext to a diverse set of other internal objects, it would be a significant undertaking to implement a session deletion feature in a way that avoids introducing use-after-free or other crashes due to referencing a BrowserContext that has been destroyed. Additionally, as you've discovered, sessions are relatively heavyweight.

**Comment by 1nfility** [comment]
Ok, so i just tested by creating 600 sessions. the memory is now at 303MB. Testing 606 proxies with 1 session and 1 request each, adds 100MB. Then, about 15 seconds later, an addional 400MB is added for about 5-10 seconds before that is released, and we are back at about 400MB.

**Comment by nornagon** [comment]
@1nfility would you be able to produce a short example that exhibits this behavior using [Electron Fiddle](/electron/fiddle)? Thanks!

**Comment by walmat** [comment]
One caveat with reusing sessions is that it doesn't seem to clear `session: true` cookies when [clearing the storage data](https://www.electronjs.org/docs/api/session#sesclearstoragedataoptions). It would be beneficial for everyone I'd assume to have at least *some* method to free up memory so that the app doesn't horde memory overtime.

I can put up a small proof of concept app that shows the inability to clear session cookies if need be, but this seems like a needed feature. If we're granted the ability to allocate memory, we should always have the power to free up that memory.

**Comment by nornagon** [comment]
@walmat I agree in principle! In practice, sessions are very heavily tied to lots of different objects in Chromium, and untangling that web to make it possible to safely garbage-collect them is a complex project. I'd be happy to review a PR, but I'm personally prioritizing other issues.

**Comment by Veiintc** [comment]
any update for session can't destroy ?
